File: Erlang/pi_parallel.erl
%  Calculation of Pi using quadrature. Parallel algorithm.
%
%  Copyright © 2008–2009, 2011, 2013  Russel Winder

-module(pi_parallel).
-export([start/0]).
-import(microsecondTime, [microsecondTime/0]).
-import(output, [out/5]).

pi_iter(_To, _From, _To, Delta, Sum) -> 4.0 * Delta * Sum ;
pi_iter( I, From, To, Delta, Sum) ->
	X =(I - 0.5) * Delta,
	pi_iter(I + 1, From, To, Delta, Sum + 1.0 /(1.0 + X * X)).

pi_iter(From, To, Delta) ->
    pi_iter(From, From, To, Delta, 0.0).

pi(N, NWorkers) ->
    NPerWorker = trunc(N / NWorkers + 0.5),
    Delta = 1 / N,
    Self = self(),
    Pids = [spawn(fun() -> Self ! pi_iter(I, lists:min([N, I + NPerWorker - 1]), Delta) end)  || I <- lists:seq(1, N, NPerWorker)],
    Results = [receive R -> R end || _ <- Pids],
    lists:sum(Results).

execute(NWorkers) ->
    N = 100000000, % 10 times fewer due to speed issues.
    StartTime = microsecondTime(),
    Pi = pi(N, NWorkers),
    ElapseTime = microsecondTime() - StartTime,
    output:out("Parallel", Pi, N, ElapseTime, NWorkers).

start() ->
    execute(1),
    io:format("~n"),
    execute(2),
    io:format("~n"),
    execute(8),
    io:format("~n"),
    execute(32),
    halt().

File: Haskell/pi_parallel.hs
--  Haskell implementation of π by Quadrature.  This implementation tries to parallelize.
--
--  Copyright © 2009–2011, 2013  Russel Winder

module Main where

import Control.Parallel (par)

import Output (outn)

piIter :: Int -> Int -> Double -> Double -> Double
piIter  n to delta accumulator
    | n > to = 4.0 * delta * accumulator
    | otherwise = piIter (n + 1) to delta (accumulator + 4.0 / (1.0 + x * x))
    where
      x = ((fromIntegral n) - 0.5) * delta

piQuadSlice :: Double -> Int -> Int -> Double
piQuadSlice delta sliceSize index = piIter start end delta 0.0
    where
      start = 1 +  index * sliceSize
      end =  (index + 1) * sliceSize

parallelMap :: (a -> b) -> [a] -> [b]
parallelMap f (x : xs) =
    let r = f x
    in r `par` r : parallelMap f xs
parallelMap _ _  = [ ]

execute :: Int -> IO()
execute numberOfSlices = do
  let n = 1000000000
  let delta = 1.0 / (fromIntegral n)
  let sliceSize = n `div` numberOfSlices
  let pi = sum (parallelMap (piQuadSlice delta sliceSize) [0 .. (numberOfSlices - 1)])
  outn "Parallel" pi n numberOfSlices

main :: IO()
main = do
  execute 1
  execute 2
  execute 8
  execute 32
